## COGS 182 Homework 4
Malachi Mabie 
2 Feb 2021

The Bellman Optimality Equations:
$$
v_*(s) = \max_a\sum_{s',r}p(s',r|s,a)\left[r+\gamma v_*(s')\right]\\
q_*(s,a) = \sum_{s',r}p(s',r|s,a)\left[r+\gamma \max_{a'}q_*(s',a')\right]\\
$$
Assuming knowledge of the environment's dynamics, this allows you to express the value of each state as a function of the values of its adjacent states. To factor policy, multiply by $\sum_a\pi(a|s)$.

### 1. (2 points) Write down the system of 14 equations and 14 unknowns for Example 4.1. 

There are four actions possible in each state of the 4x4 grid. The policy is equiprobable random.
r(s,a,s') = -1 for all states and actions. We can define 14 equations from the probabilities of their states being reached from the previous step.
$$
\begin{align}
v_*(1) &= p(1,-1|2,left)(-1 + \gamma v_*(2)) + p(1,-1|5,up)(-1+\gamma v_*(5)) + p(1,-1|1,up)(-1 + \gamma v_*(1))\\
v_*(2) &= p(2,-1|3,left)(-1 + \gamma v_*(3)) + p(2,-1|6,up)(-1+\gamma v_*(6)) + p(2,-1|1,right)(-1+\gamma v_*(1)) + p(2,-1|2,up)(-1 + \gamma v_*(2))\\
v_*(3) &= p(3,-1|3,right)(-1 + \gamma v_*(3)) + p(3,-1|7,up)(-1+\gamma v_*(7)) + p(3,-1|2,right)(-1+\gamma v_*(2)) + p(3,-1|3,up)(-1 + \gamma v_*(3))\\
v_*(4) &= p(4,-1|4,left)(-1 + \gamma v_*(4)) + p(4,-1|8,up)(-1+\gamma v_*(8)) + p(4,-1|5,left)(-1+\gamma v_*(5))\\
v_*(5) &= p(5,-1|4,right)(-1 + \gamma v_*(4)) + p(5,-1|9,up)(-1+\gamma v_*(9)) + p(5,-1|6,left)(-1+\gamma v_*(6)) + p(5,-1|1,down)(-1+\gamma v_*(1)\\
v_*(6) &= p(6,-1|5,right)(-1 + \gamma v_*(5)) + p(6,-1|10,up)(-1+\gamma v_*(10)) + p(6,-1|7,left)(-1+\gamma v_*(7)) + p(6,-1|2,down)(-1+\gamma v_*(2)\\
v_*(7) &= p(7,-1|6,right)(-1 + \gamma v_*(6)) + p(7,-1|11,up)(-1+\gamma v_*(11)) + p(7,-1|7,right)(-1+\gamma v_*(7)) + p(7,-1|3,down)(-1+\gamma v_*(3)\\
v_*(8) &= p(8,-1|8,left)(-1 + \gamma v_*(8)) + p(8,-1|12,up)(-1+\gamma v_*(12)) + p(8,-1|9,left)(-1+\gamma v_*(9)) + p(8,-1|4,down)(-1+\gamma v_*(4)\\
v_*(9) &= p(9,-1|8,right)(-1 + \gamma v_*(8)) + p(9,-1|13,up)(-1+\gamma v_*(13)) + p(9,-1|10,left)(-1+\gamma v_*(10)) + p(9,-1|5,down)(-1+\gamma v_*(5)\\
v_*(10) &= p(10,-1|9,right)(-1 + \gamma v_*(9)) + p(10,-1|14,up)(-1+\gamma v_*(14)) + p(10,-1|11,left)(-1+\gamma v_*(11)) + p(10,-1|6,down)(-1+\gamma v_*(6)\\
v_*(11) &= p(11,-1|10,right)(-1 + \gamma v_*(10)) + p(11,-1|11,right)(-1+\gamma v_*(11)) + p(11,-1|7,down)(-1+\gamma v_*(7)\\
v_*(12) &= p(12,-1|12,left)(-1 + \gamma v_*(12)) + p(12,-1|13,left)(-1+\gamma v_*(13)) + p(12,-1|8,down)(-1+\gamma v_*(8) + p(12,-1|12,down)(-1+\gamma v_*(12)\\
v_*(13) &= p(13,-1|12,right)(-1 + \gamma v_*(12)) + p(13,-1|14,left)(-1+\gamma v_*(14)) + p(13,-1|9,down)(-1+\gamma v_*(9) + p(13,-1|13,down)(-1+\gamma v_*(13)\\
v_*(14) &= p(14,-1|13,right)(-1 + \gamma v_*(13)) + p(14,-1|10,down)(-1+\gamma v_*(10) + p(14,-1|14,down)(-1+\gamma v_*(14)\\\\
\end{align}
$$
This can be simplified into the following equations:

In [ ]:
v* = -.25((cost at movement - adds to 1 or less)(cost to move - adds to 1))

v0, v15 = 0 (end positions)
v1 = .25((-1 + v1)+(-1 + v2)+(-1 + v5)+(-1 + v0))
   = -1 +.25(v0+v1+v2+v5)

In [ ]:
# 14 equations with 14 unknowns: (psuedocode)
v1 = -1 +.25(v1 + v2 + v5)
v2 = -1 +.25(v1 + v2 + v3 + v6)
v3 = -1 +.25(v2 + 2v3 + v7)
v4 = -1 +.25(v4 + v5 + v8)
v5 = -1 +.25(v1 + v4 + v6 + v9)
v6 = -1 +.25(v2 + v5 + v7 + v10)
v7 = -1 +.25(v3 + v6 + v7 + v11)
v8 = -1 +.25(v4 + v8 + v9 + v12)
v9 = -1 +.25(v5 + v8 + v10 + v13)
v10 = -1 +.25(v6 + v9 + v11 + v14)
v11 = -1 +.25(v7 + v10 + v11)
v12 = -1 +.25(v8 + 2v12 + v13)
v13 = -1 +.25(v9 + v12 + v13 + v14)
v14 = -1 +.25(v10 + v13 + v14)

### 2. (2 points) Exercise 4.1 [adjusted] if $\pi$ is the equiprobable random policy, what is $q_\pi$ (14, right)? What is $q_\pi$ (8, right)? Use the last row of Figure 4.1 as the state value function $v_\pi$ since this is what iterative policy evaluation converges to.

$$
q_\pi(14,right) = 0\\
q_\pi(8,right) = -20
$$



### 3. (2 points) What do the final values in the grid world represent? Hint: How does reward relate to steps taken?

The final values in the grid world represent the penalties incentivizing the fewest steps to terminal state.

For any initial starting position, the smallest penalizing choice is balanced relative to every next step's choice, creating an optimal path.

In [16]:
import numpy as np
import pandas as pd

pd.set_option("display.precision", 2)

# recreating original reward function
v = np.zeros(16)
display(pd.DataFrame(np.reshape(v,(-1,4))))
for i in range(140):
  vn = np.zeros(16)
  vn[1] = -1 +.25*(v[1] + v[2] + v[5])
  vn[2] = -1 + .25*(v[1] + v[2] + v[3] + v[6])
  vn[3] = -1 + .25*(v[2] + 2*v[3] + v[7])
  vn[4] = -1 + .25*(v[4] + v[5] + v[8])
  vn[5] = -1 + .25*(v[1] + v[4] + v[6] + v[9])
  vn[6] = -1 + .25*(v[2] + v[5] + v[7] + v[10])
  vn[7] = -1 + .25*(v[3] + v[6] + v[7] + v[11])
  vn[8] = -1 + .25*(v[4] + v[8] + v[9] + v[12])
  vn[9] = -1 + .25*(v[5] + v[8] + v[10] + v[13])
  vn[10] = -1 + .25*(v[6] + v[9] + v[11] + v[14])
  vn[11] = -1 + .25*(v[7] + v[10] + v[11])
  vn[12] = -1 + .25*(v[8] + 2*v[12] + v[13])
  vn[13] = -1 + .25*(v[9] + v[12] + v[13] + v[14])
  vn[14] = -1 + .25*(v[10] + v[13] + v[14])
  v = vn
display(pd.DataFrame(np.reshape(v,(-1,4))))

0    1    2    3
0  0.0  0.0  0.0  0.0
1  0.0  0.0  0.0  0.0
2  0.0  0.0  0.0  0.0
3  0.0  0.0  0.0  0.0

0     1     2     3
0   0.0 -14.0 -20.0 -22.0
1 -14.0 -18.0 -20.0 -20.0
2 -20.0 -20.0 -18.0 -14.0
3 -22.0 -20.0 -14.0   0.0

### 4. (4 points) Recreate the value function column of Figure 4.1 with an adjusted reward function. Assume that on transitions where the agent takes a down action now has a reward of -2. All other transitions have a reward of -1. For example p(9, -2 | 5, down) = 1. The policy remains equiprobable. Plot the the value function at iterations k=0,1,2,3,10, and 1000. Hint: First recreate Figure 4.1 using the original reward function.

For example, $v_*(10) = p(10,-1|9,right)(-1 + \gamma v_*(9)) + p(10,-1|14,up)(-1+\gamma v_*(14)) + p(10,-1|11,left)(-1+\gamma v_*(11)) + p(10,-2|6,down)(-2+\gamma v_*(6)$

In [22]:
import numpy as np
import pandas as pd

pd.set_option("display.precision", 2)

# reward -2 for down actions
v = np.zeros(16)
print("k = 0")
display(pd.DataFrame(np.reshape(v,(-1,4))))
for i in range(1001):
  vn = np.zeros(16)
  vn[1] = -1 + .25*(v[1] + v[2] + v[5])
  vn[2] = -1 + .25*(v[1] + v[2] + v[3] + v[6])
  vn[3] = -1 + .25*(v[2] + 2*v[3] + v[7])
  vn[4] = -1 + .25*(v[4] + v[5] + v[8])
  vn[5] = -1.25 + .25*(v[1] + v[4] + v[6] + v[9])
  vn[6] = -1.25 + .25*(v[2] + v[5] + v[7] + v[10])
  vn[7] = -1.25 + .25*(v[3] + v[6] + v[7] + v[11])
  vn[8] = -1.25 + .25*(v[4] + v[8] + v[9] + v[12])
  vn[9] = -1.25 + .25*(v[5] + v[8] + v[10] + v[13])
  vn[10] = -1.25 + .25*(v[6] + v[9] + v[11] + v[14])
  vn[11] = -1.25 + .25*(v[7] + v[10] + v[11])
  vn[12] = -1.5 + .25*(v[8] + 2*v[12] + v[13])
  vn[13] = -1.5 + .25*(v[9] + v[12] + v[13] + v[14])
  vn[14] = -1.5 + .25*(v[10] + v[13] + v[14])
  v = vn
  if i in [1,2,3,10,1000]:
    print("k = ",i)
    display(pd.DataFrame(np.reshape(v,(-1,4))))

k = 0


0    1    2    3
0  0.0  0.0  0.0  0.0
1  0.0  0.0  0.0  0.0
2  0.0  0.0  0.0  0.0
3  0.0  0.0  0.0  0.0

k =  1


0     1     2     3
0  0.00 -1.81 -2.06 -2.06
1 -1.88 -2.38 -2.44 -2.44
2 -2.50 -2.56 -2.56 -2.19
3 -2.94 -2.94 -2.56  0.00

k =  2


0     1     2     3
0  0.00 -2.56 -3.09 -3.16
1 -2.69 -3.42 -3.61 -3.53
2 -3.72 -3.84 -3.69 -3.05
3 -4.33 -4.25 -3.52  0.00

k =  3


0     1     2     3
0  0.00 -3.27 -4.11 -4.23
1 -3.46 -4.43 -4.68 -4.59
2 -4.89 -5.02 -4.75 -3.82
3 -5.66 -5.48 -4.36  0.00

k =  10


0      1      2      3
0   0.00  -7.39 -10.07 -10.80
1  -7.85 -10.02 -10.84 -10.62
2 -11.53 -11.57 -10.44  -8.05
3 -13.18 -12.21  -8.88   0.00

k =  1000


0      1      2      3
0   0.00 -16.53 -23.54 -25.82
1 -17.13 -22.04 -24.29 -24.10
2 -25.34 -25.21 -22.46 -17.19
3 -28.68 -26.02 -18.16   0.00

### 5.  (up to 4 extra points) Exercise 4.9 [adjusted] Implement value iteration for the gambler’s problem Example 4.3 and solve it for ph =0.3. In programming, you may find it convenient to introduce two dummy states corresponding to termination with capital of 0 and 100, giving them values of 0 and 1 respectively. Show your results graphically, as in Figure 4.3.